### PR TITLE
WIP Pilot client mcp integration Phase 2

### PR DESCRIPTION
**Pull request by nhsieh**
[DO NOT MERGE] 

This PR is the phase two (ServiceDiscovery) of the MCP client integration work.

### CODE REVIEW

**Comment by costinm** [comment]
/hold

Service discovery - endpoints in particular - have different requirements and AFAIK are not supposed to
be handled by Galley. Pilot has a different interface and behavior for endpoints - and any replacement
will need to take into account the scalability and load balancing needs.

Can we finish testing config decomposition - and not mix it up with the refactoring of Endpoints and EDS ?

**Comment by costinm** [comment]
Also: I think we repeatedly discussed and agreed that for endpoints we'll use directly EDS - not MCP.
Endpoints include dynamic information - like health status, support for lame duck/draining, locality, source-based subsetting, split horizon, etc - and very different scalability and change needs.  
I don't think the design for MCP covers those aspects - and this conflicts with a lot of the other work we are doing on endpoint scalability.

**Comment by Nino-K** [comment]
@costinm 
> Can we finish testing config decomposition - and not mix it up with the refactoring of Endpoints and EDS ?

Config decomposition was phase one of this MCP PR https://github.com/istio/istio/pull/7631/ as part of the PR we provided and e2e test and additional tests for the MCP controller, is there anything else you have in mind?

**Comment by costinm** [comment]
My point is that I would prefer the config decomposition is finished -
including scalability testing, prod, etc - before starting work on
endopoints, which
have very different characteristics.

It is not the same thing - and we are actively working on some
endpoint-related projects that may overlap with this change.

1.1 is still far away - any reason to hurry up with this PR now ? I want to
make sure the implementation is aligned with the plans for
zero vpn and scalability of endpoints.

File: pilot/pkg/bootstrap/server.go [hunk 1]
```diff
@@ -755,12 +761,23 @@ func (s *Server) initConfigRegistry(serviceControllers *aggregate.Controller) {
...
 
 func (s *Server) initDiscoveryService(args *PilotArgs) error {
-	environment := &model.Environment{
-		Mesh:             s.mesh,
-		IstioConfigStore: s.istioConfigStore,
-		ServiceDiscovery: s.ServiceController,
-		ServiceAccounts:  s.ServiceController,
-		MixerSAN:         s.mixerSAN,
+	var environment *model.Environment
+	if mcpEnabled(args) {
```
Comment (costinm) [inline pilot/pkg/bootstrap/server.go]: mcp enabled means it'll also fetch from mcp - I assume the other controlers will still be enabled.

**Comment by costinm** [comment]
In particular: the representation of endpoints and the interface used for endpoint discovery in pilot are extremely likely to change - both to support split horizon, and to allow pushing only modified clusters ( which is critical for scalability ). And the changes will happen in 1.0 branch ( since perf/scale optimization goes there ) - this PR would make it very hard to merge.

**Comment by andraxylia** [comment]
@costinm large design changes that you are talking about, or features like Split Horizon EDS design for multi-cluster cannot go in release 1.0 and invalidate all the work that people are doing in master.  I added this for discussion in the TOC meeting. 

Do you have a design doc for the changes in service discovery? I will submit a PR in istio/api with the new Services and Endpoints, and link it here. We need to find alignment, I am sure there is a solution.

cc @linsun

**Comment by Nino-K** [comment]
@costinm with pilot decomposition document in mind and having to deliver the first phase (IstioConfigStore) we were under the assumption that serviceDiscovery/ endpoints are going to be next to tackle. However, with all that in mind what do you think makes the most sense to deliver next? Should serviceDiscovery work be put on hold until we figure out the interface changes for endpoints/serviceDiscovery? It would be great to get some clarity for what should be done next.

**Comment by costinm** [comment]
I think it would be better to finish up IstioConfigStore before starting on endpoints, and also to wait on endpoints for at least few weeks. 

The "Decomposition document" is a design doc - which needs to be implemented and proven. We have many approved docs - not all will work in practice as well as it is claimed. 

The current interfaces for endpoints are clearly in need of change - and we expect at least 2 major changes in 1.0 plus a smaller one in master. The interface for endpoints appears to be again up to debate - and I suspect will be a very lively debate if anyone is proposing a 5th representation for endpoints ( in Istio we have K8S Endpoint, our ServiceEndpoint, Envoy's endpoint - plus the ServiceInstance interface used all over the code).  On top of this - we have the on-demand cluster
proposal that will certainly have an impact on how pilot handles endpoints and clusters. 

Config handling in Galley is not only about MCP - one of the reasons to have galley is to provide 
validations beyond what Pilot already does, and as any new component it will need a _lot_ of work 
before it can be used in production. Including tests to compare it against the current implementation
(which probably will happen from 1.1 to 1.2, while both will be available). There is no shortage of work to be done to finish up config.

On the goal of removing the K8S dependency - service discovery already has 3-4 non-k8s registries and with config using MCP it is already possible to use Pilot without K8S.  So anyone who needs no-k8s is not blocked by the endpoint work.

**Comment by andraxylia** [comment]
@costinm you are commenting on the PR that is part of implementing the design doc :)

**Comment by andraxylia** [comment]
Here is  [WIP] API proposal for Services and Endpoints, that attempts to re-use the Envoy endpoint, which I suppose is your main concern : https://github.com/istio/api/pull/628

You can jump fiercely on that one and leave this one alone :).

**Comment by costinm** [comment]
Until the API and the new model for endpoint handling  are clear - I think
everyone should leave this PR alone and not waste more time.

**Pull request closed without merging**